external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala

```diff
@@ -0,0 +1,344 @@
+      seekToLatest: Boolean): Map[TopicPartition, Long] = withRetries {
+
+    // Poll to get the latest assigned partitions
+    logTrace("\tPolling")
+    consumer.poll(0)
+    val partitions = consumer.assignment()
+    consumer.pause(partitions)
+    logDebug(s"\tPartitioned assigned to consumer: $partitions")
+
+    // Get the current or latest offset of each partition
+    if (seekToLatest) {
+      consumer.seekToEnd(partitions)
+      logDebug("\tSeeked to the end")
+    }
+    logTrace("Getting positions")
+    val partitionOffsets = partitions.asScala.map(p => p -> consumer.position(p)).toMap
+    logDebug(s"Got offsets for partition : $partitionOffsets")
+    partitionOffsets
+  }
+
+  /** Fetch the earliest offsets for newly discovered partitions */
+  private def fetchNewPartitionEarliestOffsets(
+      newPartitions: Seq[TopicPartition]): Map[TopicPartition, Long] = withRetries {
+
+    // Poll to get the latest assigned partitions
+    logTrace("\tPolling")
+    consumer.poll(0)
+    val partitions = consumer.assignment()
+    logDebug(s"\tPartitioned assigned to consumer: $partitions")
+    require(newPartitions.forall(tp => partitions.contains(tp)),
+      s"$partitions doesn't contain all new paritions: $newPartitions")
+
+    // Get the earliest offset of each partition
+    consumer.seekToBeginning(newPartitions.asJava)
+    val partitionToOffsets = newPartitions.map(p => p -> consumer.position(p)).toMap
+    logDebug(s"Got offsets for new partitions: $partitionToOffsets")
+    partitionToOffsets
+  }
+
+  /**
+   * Helper function that does multiple retries on the a body of code that returns offsets.
+   * Retries are needed to handle transient failures. For e.g. race conditions between getting
+   * assignment and getting position while topics/partitions are deleted can cause NPEs.
+   */
+  private def withRetries(
+      body: => Map[TopicPartition, Long]): Map[TopicPartition, Long] = synchronized {
+
+    var result: Option[Map[TopicPartition, Long]] = None
+    var attempt = 1
+    var lastException: Exception = null
+    while (result.isEmpty && attempt <= MAX_OFFSET_FETCH_ATTEMPTS) {
+      try {
+        result = Some(body)
+      } catch {
+        case e: Exception =>
+          lastException = e
+          logWarning(s"Error in attempt $attempt getting Kafka offsets: ", e)
+          attempt += 1
+          Thread.sleep(OFFSET_FETCH_ATTEMPT_INTERVAL_MS)
+      }
+    }
+    if (result.isEmpty) {
+      assert(attempt > MAX_OFFSET_FETCH_ATTEMPTS)
+      assert(lastException != null)
+      throw lastException
+    }
+    result.get
+  }
+
+  private def positiveMod(a: Long, b: Int): Int = ((a % b).toInt + b) % b
+}
+
+
+/** Companion object for the [[KafkaSource]]. */
+private[kafka010] object KafkaSource {
+
+  val MAX_OFFSET_FETCH_ATTEMPTS = 3
+  val OFFSET_FETCH_ATTEMPT_INTERVAL_MS = 10
+
+  def kafkaSchema: StructType = StructType(Seq(
+    StructField("checksum", LongType),
```

Comment:
Unless you're 100% sure what people are doing with custom deserializers, I'd recommend keeping everything available that could be obtained from a ConsumerRecord.  If you're concerned about bloat, give people a choice between 2 schema, one with all the fields and one with just key / value.